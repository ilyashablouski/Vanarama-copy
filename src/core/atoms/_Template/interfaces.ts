import { IBaseProps } from '../../interfaces/base';

export interface ITemplateProps extends IBaseProps {}
