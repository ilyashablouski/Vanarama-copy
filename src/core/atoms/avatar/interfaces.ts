import { IBaseProps } from '../../interfaces/base';
import { TSize } from '../../../types/size';

export interface IAvatarProps extends IBaseProps {
  src?: string;
  size?: TSize;
  altText?: string;
}
