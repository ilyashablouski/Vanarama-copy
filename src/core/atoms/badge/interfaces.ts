import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IBadgeProps extends IBaseProps {
  size?: TSize;
  color?: TColor;
  label?: string;
}
