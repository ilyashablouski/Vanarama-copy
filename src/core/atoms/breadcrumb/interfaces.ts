import { IBaseProps } from '../../interfaces/base';
import { ILink } from '../../interfaces/link';

export interface IBreadcrumbProps extends IBaseProps {
  items?: ILink[];
}
