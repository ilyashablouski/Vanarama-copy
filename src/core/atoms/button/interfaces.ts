import { ButtonHTMLAttributes } from 'react';
import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IButtonProps
  extends IBaseProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  color?: TColor;
  fill?: 'solid' | 'outline' | 'clear';
  icon?: React.ReactNode;
  iconColor?: TColor;
  iconPosition?: 'before' | 'after';
  label?: string | React.ReactNode;
  round?: boolean;
  size?: TSize;
  withoutDefaultClass?: boolean;
  customCTAColor?: string;
}
