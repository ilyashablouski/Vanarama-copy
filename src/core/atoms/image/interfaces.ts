import { ImageProps } from 'next/image';

import { IBaseProps } from '../../interfaces/base';
import { TSize } from '../../../types/size';

import { ICloudflareOptimisation } from '../../../helpers/imageOptimiseUtils/interfaces';

export interface IImageProps extends IBaseProps {
  src: string;
  optimisedHost?: string;
  optimisationOptions?: ICloudflareOptimisation;
  alt?: string;
  size?: TSize;
  round?: boolean;
  plain?: boolean;
  inline?: boolean;
  width?: string | number;
  height?: string | number;
  lazyLoad?: boolean;
  dataTestId?: string;
}

export interface IImageV2Props extends IBaseProps {
  width?: string | number;
  height?: string | number;
  optimisedHost?: string;
  optimisationOptions?: ICloudflareOptimisation;
  src: string;
  alt?: string;
  size?: TSize;
  round?: boolean;
  inline?: boolean;
  lazyLoad?: boolean;
  objectFit?: ImageProps['objectFit'];
  plain?: boolean;
}
