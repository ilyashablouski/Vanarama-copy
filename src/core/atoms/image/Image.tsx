import React, { FC } from 'react';
import cx from 'classnames';

import { IImageProps } from './interfaces';

import {
  DEFAULT_OPTIMISATION,
  optimiseImage,
} from '../../../helpers/imageOptimiseUtils';

const Image: FC<IImageProps> = props => {
  const {
    className,
    size = 'expand',
    width,
    height,
    alt,
    round,
    plain,
    inline,
    onError,
    optimisedHost,
    optimisationOptions,
  } = props;

  const { src } = props;

  let srcset;
  let srcDefault = src;
  let src1200;
  let src800;
  let src320;

  // Check if image should be optimised.
  if (optimisedHost) {
    srcDefault = optimiseImage(optimisedHost, src, {
      ...DEFAULT_OPTIMISATION,
      ...optimisationOptions,
    });
    src1200 = optimiseImage(optimisedHost, src, {
      ...DEFAULT_OPTIMISATION,
      width: 1200,
      height: 1200,
      ...optimisationOptions,
    });
    src800 = optimiseImage(optimisedHost, src, {
      ...DEFAULT_OPTIMISATION,
      width: 800,
      height: 800,
      quality: 70,
      ...optimisationOptions,
    });
    src320 = optimiseImage(optimisedHost, src, {
      ...DEFAULT_OPTIMISATION,
      width: 320,
      height: 320,
      quality: 60,
      ...optimisationOptions,
    });

    srcset = `${src320} 320w, ${src800} 800w, ${src1200} 1200w`;
  }

  return (
    <div
      className={cx('image', {
        [`${className}`]: className,
        [`-${size}`]: size,
        '-round': round,
        '-plain': plain,
        '-inline': inline,
      })}
    >
      <img
        loading="lazy"
        srcSet={srcset}
        sizes="(min-width:3200px) 800px, 1200px"
        alt={alt}
        width={width}
        height={height}
        className="image--native"
        src={srcDefault || src}
        onError={onError}
      />
    </div>
  );
};

Image.displayName = 'Image';

export default React.memo(Image);
