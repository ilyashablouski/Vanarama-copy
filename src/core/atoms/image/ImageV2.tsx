import React, { FC, SyntheticEvent } from 'react';
import Image from 'next/image';
import cx from 'classnames';

import { optimizedLoader } from './helpers';
import { IImageV2Props } from './interfaces';

const PLACEHOLDER_URL = `${process.env.HOST_DOMAIN}/vehiclePlaceholder.jpg`;

const ImageV2: FC<IImageV2Props> = ({
  className,
  size = 'expand',
  width,
  height,
  alt,
  round,
  plain,
  inline,
  dataTestId,
  objectFit,
  optimisedHost,
  sizes,
  quality,
  lazyLoad,
  dataUiTestId,
  ...props
}) => {
  let { src } = props;

  src = src || PLACEHOLDER_URL;
  if (src.search(/^http[s]?:/) === -1) {
    src = `https:${src}`;
  }

  const onError = (e: SyntheticEvent<HTMLImageElement>) => {
    e.currentTarget.srcset = '';
    e.currentTarget.src = PLACEHOLDER_URL;
  };

  const layout = width && height ? 'responsive' : 'fill';
  const loader = optimisedHost ? optimizedLoader : undefined;

  return (
    <div
      className={cx('image', {
        [`${className}`]: className,
        [`-${size}`]: size,
        '-round': round,
        '-plain': plain,
        '-inline': inline,
      })}
      data-uitestid={dataUiTestId ? `${dataUiTestId}_img-wrapper` : undefined}
    >
      <Image
        alt={alt}
        width={width}
        height={height}
        src={src}
        layout={layout}
        loader={loader}
        sizes={sizes}
        quality={quality}
        data-testid={dataTestId}
        className="image--native"
        objectFit={objectFit}
        priority={lazyLoad === false}
        onError={onError}
        data-uitestid={dataUiTestId ? `${dataUiTestId}_img` : undefined}
      />
    </div>
  );
};

ImageV2.displayName = 'ImageV2';

export default React.memo(ImageV2);
