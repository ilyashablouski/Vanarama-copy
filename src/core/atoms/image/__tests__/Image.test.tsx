import React from 'react';
import renderer from 'react-test-renderer';

import ImageV2 from 'core/atoms/image/ImageV2';
import Image from '..';
import { IImageProps, IImageV2Props } from '../interfaces';

const optionalProps = {
  id: 'component',
  className: 'class',
  size: 'small' as const,
  color: 'orange' as const,
};

const mandatoryProps = {
  src: 'https://source.unsplash.com/collection/2102317/1000x650?sig=40349',
};

function getComponentImage(props?: Partial<IImageProps>) {
  return renderer
    .create(<Image src={mandatoryProps.src} {...props} />)
    .toJSON();
}

function getComponentImageV2(props?: Partial<IImageV2Props>) {
  return renderer
    .create(<ImageV2 src={mandatoryProps.src} {...props} />)
    .toJSON();
}

describe('<Image />', () => {
  it('renders Image correctly', () => {
    const tree = getComponentImage();
    expect(tree).toMatchSnapshot();
  });

  it('renders Image correctly with optional props', () => {
    const tree = getComponentImage(optionalProps);
    expect(tree).toMatchSnapshot();
  });
});

describe('<ImageV2 />', () => {
  it('renders ImageV2 correctly', () => {
    const tree = getComponentImageV2();
    expect(tree).toMatchSnapshot();
  });

  it('renders ImageV2 correctly with optional props', () => {
    const tree = getComponentImageV2(optionalProps);
    expect(tree).toMatchSnapshot();
  });
});
