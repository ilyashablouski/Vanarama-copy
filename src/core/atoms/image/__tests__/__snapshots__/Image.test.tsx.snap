// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> renders Image correctly 1`] = `
<div
  className="image -expand"
>
  <img
    className="image--native"
    data-cfasync="false"
    loading="eager"
    onError={[Function]}
    sizes="(min-width:320px) 800px, 1200px"
    src="https://source.unsplash.com/collection/2102317/1000x650?sig=40349"
  />
</div>
`;

exports[`<Image /> renders Image correctly with optional props 1`] = `
<div
  className="image class -small"
>
  <img
    className="image--native"
    data-cfasync="false"
    loading="eager"
    onError={[Function]}
    sizes="(min-width:320px) 800px, 1200px"
    src="https://source.unsplash.com/collection/2102317/1000x650?sig=40349"
  />
</div>
`;

exports[`<ImageV2 /> renders ImageV2 correctly 1`] = `
<div
  className="image -expand"
>
  <span
    style={
      Object {
        "background": "none",
        "border": 0,
        "bottom": 0,
        "boxSizing": "border-box",
        "display": "block",
        "height": "initial",
        "left": 0,
        "margin": 0,
        "opacity": 1,
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": "initial",
      }
    }
  >
    <img
      className="image--native"
      data-nimg="fill"
      decoding="async"
      onError={[Function]}
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style={
        Object {
          "border": "none",
          "bottom": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": 0,
          "left": 0,
          "margin": "auto",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "100%",
          "minWidth": "100%",
          "objectFit": undefined,
          "objectPosition": undefined,
          "padding": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 0,
        }
      }
    />
    <noscript>
      <img
        className="image--native"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        onError={[Function]}
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=3840&q=75"
        srcSet="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=3840&q=75 3840w"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
    </noscript>
  </span>
</div>
`;

exports[`<ImageV2 /> renders ImageV2 correctly with optional props 1`] = `
<div
  className="image class -small"
>
  <span
    style={
      Object {
        "background": "none",
        "border": 0,
        "bottom": 0,
        "boxSizing": "border-box",
        "display": "block",
        "height": "initial",
        "left": 0,
        "margin": 0,
        "opacity": 1,
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": "initial",
      }
    }
  >
    <img
      className="image--native"
      data-nimg="fill"
      decoding="async"
      onError={[Function]}
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style={
        Object {
          "border": "none",
          "bottom": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": 0,
          "left": 0,
          "margin": "auto",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "100%",
          "minWidth": "100%",
          "objectFit": undefined,
          "objectPosition": undefined,
          "padding": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 0,
        }
      }
    />
    <noscript>
      <img
        className="image--native"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        onError={[Function]}
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=3840&q=75"
        srcSet="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fcollection%2F2102317%2F1000x650%3Fsig%3D40349&w=3840&q=75 3840w"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
    </noscript>
  </span>
</div>
`;
